modules/storages/app/contracts/storages/file_links/create_contract.rb

```diff
@@ -0,0 +1,30 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2021 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class Storages::FileLinks::CreateContract < BaseContract
```

Comment:
It is the `BaseContract` from `app/contracts/base_contract.rb`. It was the minimal amount of code needed to make the test pass. It is used up by the `Storages::FileLinks::CreateService` (from `BaseContracted#validate_contract`) and deduced in `API::V3::FileLinks::CreateEndpoint` (from API::Utilities::Endpoints::Bodied#deduce_process_contract`).

Reply:
It annoys me that model validation happens _after_ contract validation. It would more sense to validate the model first (smallest "size"/most basic) and then validate the contract, so that the simple validations like `validates :presence` can belong to the model instead of being moved to the contract. 

It's only when a validation is valid for a certain context (in a service and not another) that it can be interesting to move the model validation to specific contracts.

Or another approach could be to move validations in a concern, and include them in both the contract and the model.

Reply:
I agree, that I seems more intuitive to have the model validations run first. @ulferts?

Another question is also where the specs for which validations go if validations are spread across models and contracts. It increases spec code.

Reply:
Yes, and I looked into it, and the benefits were not that obvious, so I stayed on the safe side and decided to not touch it and let it like this. But I'll keep it in mind for the future.